lsbi/network.py

```diff
@@ -0,0 +1,202 @@
+        self.batch_norm_2 = nn.BatchNorm1d(internal_dim)
+        self.dense_3 = nn.Linear(internal_dim, internal_dim)
+        self.leaky_relu_3 = nn.LeakyReLU()
+        self.batch_norm_3 = nn.BatchNorm1d(internal_dim)
+        self.dense_4 = nn.Linear(internal_dim, internal_dim)
+        self.leaky_relu_4 = nn.LeakyReLU()
+        self.dense_5 = nn.Linear(internal_dim, 1)
+
+    def forward(self, x):
+        """Forward pass through the network, logit output."""
+        x = self.dense(x)
+        x = self.leaky_relu(x)
+        x = self.batch_norm(x)
+        x = self.dense_1(x)
+        x = self.leaky_relu_1(x)
+        x = self.batch_norm_1(x)
+        x = self.dense_2(x)
+        x = self.leaky_relu_2(x)
+        x = self.batch_norm_2(x)
+        x = self.dense_3(x)
+        x = self.leaky_relu_3(x)
+        x = self.batch_norm_3(x)
+        x = self.dense_4(x)
+        x = self.leaky_relu_4(x)
+        x = self.dense_5(x)
+        return x
+
+    def loss(self, x):
+        """Loss function for the network."""
+        raise NotImplementedError
+
+    def predict(self, x):
+        """Predict the Bayes Factor."""
+        raise NotImplementedError
+
+
+class BinaryClassifier(BinaryClassifierBase):
+    """
+    Extends the BinaryClassifierBase to use a BCE loss function.
+
+    Furnishes with a direction prediction of the Bayes Factor.
+    """
+
+    def loss(self, x, target):
+        """Binary cross entropy loss function for the network."""
+        x = self.forward(x)
+        return nn.BCEWithLogitsLoss()(x, target)
+
+    def predict(self, x):
+        """Predict the log Bayes Factor.
+
+        log K = lnP(Class 1) - lnP(Class 0)
+        """
+        x = torch.tensor(x, dtype=torch.float32)
+        pred = self.forward(x)
+        pred = nn.Sigmoid()(pred)
+        return (torch.log(pred) - torch.log(1-pred)).detach().numpy()
+
+
+class BinaryClassifierLPop(BinaryClassifierBase):
+    """
+    Extends the BinaryClassifierBase to use a LPop Exponential loss.
+
+    Furnishes with a direction prediction of the Bayes Factor.
+    """
+
+    def lpop(self, x, alpha=2.0):
+        """Leaky parity odd power transform."""
+        return x + x * torch.pow(torch.abs(x), alpha - 1.0)
+
+    def loss(self, x, target, alpha=2.0):
+        """Lpop Loss function for the network."""
+        x = self.forward(x)
+        return torch.exp(
+            torch.logsumexp((0.5 - target) * self.lpop(x, alpha=alpha), dim=0)
+            - torch.log(torch.tensor(x.shape[0], dtype=torch.float64))
+        ).squeeze()
+
+    def predict(self, x, alpha=2.0):
+        """Predict the log Bayes Factor.
+
+        log K = lnP(Class 1) - lnP(Class 0)
+        """
+        x = torch.tensor(x, dtype=torch.float32)
+        pred = self.forward(x)
+        pred = self.lpop(pred, alpha=alpha)
+        return pred.detach().numpy()
+
+
+def train(
+    model,
+    data,
+    labels,
+    num_epochs=10,
+    batch_size=128,
+    decay_rate=0.95,
+    lr=0.001,
+):
```

Comment:
I know a separate 'train' function is how it is done in the deep learning/torch literature, but at this point we could consider making this a scikitlearn-like train/predict member function. At some point (not this PR), we could even wrap in sklearn architecture.

Reply:
agreed, especially as the loss etc. is folded into the classifier it makes more sense in future to just reject the functional ml workflow in favour of oo

Reply:
Are we going to do this in this PR?

Reply:
moved train to the model, now called as `fit(X,y)` a la sklearn